### PR TITLE
Fix #211: include non-html filenames in page_ids

### LINKED ISSUE #211: Bare .haml extension not recognised by HighVoltage.page_ids
The `haml-rails` gem (as opposed to `haml`) allows one to name its views `foo.haml` instead of the more cumbersome `foo.html.haml`.

However HighVoltage.page_ids does not list shortcutted pages, even if they are rendered when one visits `/pages/foo`.

In other words: high_voltage plays fine with haml-rails except for shortcutted names which aren't included in `HighVoltage.page_ids`.

Could you please address this?

Thanks!

Victor

### PR DESCRIPTION
**Pull request by jonathanhefner**
This refactoring adds consideration for filenames without a ".html" extension when listing all page_ids (e.g. when generating a sitemap).  For example, when using the Haml gem, a file named "template.haml" will now cause "template" to appear in `HighVoltage.page_ids`. (Previously, this required the file have a ".html.haml" extension, even though that was not required by the Haml gem itself.)

This also removes the assumption that template filename extensions consist of only lower-case alpha characters.  For example, if someone decides to go wild and create a template engine based on the [m4 macro language](https://en.wikipedia.org/wiki/M4_%28computer_language%29), "template.html.m4" would now appear as "template" in `HighVoltage.page_ids`.

This also now allows (static or templated) plain text files to appear in `HighVoltage.page_ids` with their extension intact.  For example, "foo.txt.erb" and "bar.txt" will appear as "foo.txt" and "bar.txt".

This also reduces the number of disk accesses when listing page_ids.

### CODE REVIEW

File: spec/high_voltage/page_collector_spec.rb [hunk 1]
```diff
@@ -1,15 +1,30 @@
...
+  before :example do
+    # For whatever reason, the HTML template handler is not available at
+    # this point, causing the ".html" file extension to be unrecognized
+    # by our code.  Thus we plug the ERB handler in its place.
+    ActionView::Template.register_template_handler(:html,
+      ActionView::Template.handler_for_extension(:erb))
+
+    # Also, pretend we have the Haml gem installed...
+    ActionView::Template.register_template_handler(:haml,
+      ActionView::Template.handler_for_extension(:raw))
```
Comment (jonathanhefner) [inline spec/high_voltage/page_collector_spec.rb]: Aligning the parameters would violate the 80-character line limit...

File: spec/high_voltage/page_collector_spec.rb [hunk 2]
```diff
@@ -1,15 +1,30 @@
 require "spec_helper"
 
 describe HighVoltage::PageCollector do
+  before :example do
+    # For whatever reason, the HTML template handler is not available at
+    # this point, causing the ".html" file extension to be unrecognized
+    # by our code.  Thus we plug the ERB handler in its place.
+    ActionView::Template.register_template_handler(:html,
+      ActionView::Template.handler_for_extension(:erb))
```
Comment (jonathanhefner) [inline spec/high_voltage/page_collector_spec.rb]: Aligning the parameters would violate the 80-character line limit...

**Comment by matthewford** [comment]
👍

**Comment by dgalarza** [comment]
Thanks @jonathanhefner - closing this in favor of https://github.com/thoughtbot/high_voltage/pull/266 which keeps `Page` intact but utilizes `ActionView::Template.template_handler_extensions` as introduced here.

**Comment by jonathanhefner** [comment]
(Reposting https://github.com/thoughtbot/high_voltage/pull/266#issuecomment-379351817 here, for posterity)

The reason I took this approach was that it resulted in less complexity (fewer moving parts, fewer lines of code).  It was also more performant (fewer object allocations and fewer system calls per request).  Also the name `Page` seemed to be a misnomer, because `Page` objects could represent file paths which were explicitly not pages.  It seemed like the only purpose of the `Page` class was to obfuscate filtering logic that was part of `PageCollector`.

**Pull request closed without merging**